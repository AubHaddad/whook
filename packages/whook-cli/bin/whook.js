const path = require('path');
const run = require('../dist/index').default;
const { prepareEnvironment } = require(path.join(
  process.cwd(),
  'dist',
  'index',
));

run({ prepareEnvironment });
