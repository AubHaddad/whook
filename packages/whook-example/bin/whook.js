#! /usr/bin/env node

import path from 'path';
import runCLI from '@whook/whook';

const { prepareEnvironment } = await import(
  path.join(process.cwd(), 'src', 'index.js')
);

await runCLI(prepareEnvironment);
