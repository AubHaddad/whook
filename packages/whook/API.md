# API
## Functions

<dl>
<dt><a href="#prepareServer">prepareServer(injectedNames, $)</a> ⇒</dt>
<dd><p>Runs the Whook server</p>
</dd>
<dt><a href="#prepareEnvironment">prepareEnvironment($)</a> ⇒</dt>
<dd><p>Prepare the Whook server environment</p>
</dd>
<dt><a href="#initCONFIGS">initCONFIGS(services, [log])</a> ⇒ <code>Promise.&lt;String&gt;</code></dt>
<dd><p>Initialize the CONFIGS serviceaccording to the NODE_ENV</p>
</dd>
<dt><a href="#initENV">initENV(services, [log])</a> ⇒ <code>Promise.&lt;Object&gt;</code></dt>
<dd><p>Initialize the ENV service using process env plus dotenv files</p>
</dd>
<dt><a href="#initHOST">initHOST(services, [log])</a> ⇒ <code>Promise.&lt;String&gt;</code></dt>
<dd><p>Initialize the HOST service from ENV or auto-detection if
 none specified in ENV</p>
</dd>
<dt><a href="#initPORT">initPORT(services, [log])</a> ⇒ <code>Promise.&lt;Number&gt;</code></dt>
<dd><p>Initialize the PORT service from ENV or auto-detection if
 none specified in ENV</p>
</dd>
<dt><a href="#initProjectDir">initProjectDir(services, [log])</a> ⇒ <code>Promise.&lt;string&gt;</code></dt>
<dd><p>Auto detect the Whook PROJECT_DIR</p>
</dd>
</dl>

<a name="prepareServer"></a>

## prepareServer(injectedNames, $) ⇒
Runs the Whook server

**Kind**: global function  
**Returns**: Object
A promise of the injected services  

| Param | Type | Description |
| --- | --- | --- |
| injectedNames | <code>Array.&lt;String&gt;</code> | Root dependencies names to instanciate and return |
| $ | <code>Knifecycle</code> | The Knifecycle instance to use for the server run |

<a name="prepareEnvironment"></a>

## prepareEnvironment($) ⇒
Prepare the Whook server environment

**Kind**: global function  
**Returns**: Promise<Knifecycle>
A promise of the Knifecycle instance  

| Param | Type | Description |
| --- | --- | --- |
| $ | <code>Knifecycle</code> | The Knifecycle instance to set the various services |

<a name="initCONFIGS"></a>

## initCONFIGS(services, [log]) ⇒ <code>Promise.&lt;String&gt;</code>
Initialize the CONFIGS serviceaccording to the NODE_ENV

**Kind**: global function  
**Returns**: <code>Promise.&lt;String&gt;</code> - A promise of a containing the actual host.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| services | <code>Object</code> |  | The services CONFIGS depends on |
| services.NODE_ENV | <code>Object</code> |  | The injected NODE_ENV value |
| services.PROJECT_SRC | <code>Object</code> |  | The project source directory |
| [log] | <code>Object</code> | <code>noop</code> | An optional logging service |

<a name="initENV"></a>

## initENV(services, [log]) ⇒ <code>Promise.&lt;Object&gt;</code>
Initialize the ENV service using process env plus dotenv files

**Kind**: global function  
**Returns**: <code>Promise.&lt;Object&gt;</code> - A promise of an object containing the actual env vars.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| services | <code>Object</code> |  | The services ENV depends on |
| services.NODE_ENV | <code>Object</code> |  | The injected NODE_ENV value to lookk for `.env.${NODE_ENV}` env file |
| services.PWD | <code>Object</code> |  | The process current working directory |
| [services.BASE_ENV] | <code>Object</code> | <code>{}</code> | An optional base environment |
| [log] | <code>Object</code> | <code>noop</code> | An optional logging service |

<a name="initHOST"></a>

## initHOST(services, [log]) ⇒ <code>Promise.&lt;String&gt;</code>
Initialize the HOST service from ENV or auto-detection if
 none specified in ENV

**Kind**: global function  
**Returns**: <code>Promise.&lt;String&gt;</code> - A promise of a containing the actual host.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| services | <code>Object</code> |  | The services HOST depends on |
| [services.ENV] | <code>Object</code> | <code>{}</code> | An optional environment object |
| [log] | <code>Object</code> | <code>noop</code> | An optional logging service |

<a name="initPORT"></a>

## initPORT(services, [log]) ⇒ <code>Promise.&lt;Number&gt;</code>
Initialize the PORT service from ENV or auto-detection if
 none specified in ENV

**Kind**: global function  
**Returns**: <code>Promise.&lt;Number&gt;</code> - A promise of a number representing the actual port.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| services | <code>Object</code> |  | The services PORT depends on |
| [services.ENV] | <code>Object</code> | <code>{}</code> | An optional environment object |
| [log] | <code>Object</code> | <code>noop</code> | An optional logging service |

<a name="initProjectDir"></a>

## initProjectDir(services, [log]) ⇒ <code>Promise.&lt;string&gt;</code>
Auto detect the Whook PROJECT_DIR

**Kind**: global function  
**Returns**: <code>Promise.&lt;string&gt;</code> - A promise of a number representing the actual port.  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| services | <code>Object</code> |  | The services PROJECT_DIR depends on |
| services.PWD | <code>Object</code> |  | The process working directory |
| [log] | <code>Object</code> | <code>noop</code> | An optional logging service |

