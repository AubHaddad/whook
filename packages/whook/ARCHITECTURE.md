[//]: # ( )
[//]: # (This file is automatically generated by the `jsarch`)
[//]: # (module. Do not change it elsewhere, changes would)
[//]: # (be overriden.)
[//]: # ( )
# Architecture Notes



## Server run

Whook exposes a `runServer` function to programmatically spawn
 its server. It is intended to be reusable and injectable so
 that projects can override the whole `whook` default behavior.

[See in context](./src/index.js#L21-L25)



## Server preparation

Whook exposes a `prepareServer` function to create its server
 configuration. It takes eventually additional injections that
 would be required at a higher level and a
 [Knifecycle](https://github.com/nfroidure/knifecycle)
 containing the bootstrapped environment and allowing
 to complete and run the server.

[See in context](./src/index.js#L50-L57)



### Root injections

* We need to inject `httpServer` and `process` to bring life to our
 *  server. We also inject `log` for logging purpose and custom other
 *  injected name that were required upfront.

[See in context](./src/index.js#L68-L72)



## Server environment

The Whook `prepareEnvironment` function aims to provide the complete
 server environment without effectively planning its run. It allows
 to use that environment for CLI or build purposes. It also
 provides a chance to override some services/constants
 before actually preparing the server.

[See in context](./src/index.js#L82-L88)



### `PWD` env var

The Whook server heavily rely on the process working directory
 to dynamically load contents. We are making it available to
 the DI system as a constant.

[See in context](./src/index.js#L97-L101)



### `NODE_ENV` env var

Whook has different behaviors depending on the `NODE_ENV` value
 consider setting it to production before shipping.

[See in context](./src/index.js#L105-L108)



### `WHOOK_SRC` env var

Whook need to know its main directories to be consistent
 between builds and development environments.

[See in context](./src/index.js#L112-L115)



### Logging

Whook's default logger write to the NodeJS default console
 except for debugging messages where it use the `debug`
 module so that you can set the `DEBUG` environment
 variable to `whook` and get debug messages in output.

[See in context](./src/index.js#L118-L123)



### Initializers

Whook's embed a few default initializers proxied from
 `common-services`, `swagger-http-router` or it own
 `src/services` folder. It can be wrapped or overriden
 at will later in project's main file.

[See in context](./src/index.js#L137-L142)



## Environment service

The `ENV` service add a layer of configuration over just using
 node's `process.env` value. Beware that `PWD` and `NODE_ENV` are
 guaranteed to be the exact same than the injected constants.
 It is up to you to decide upstream if you set them via the
 `process.env.NODE_ENV` and `process.cwd()` values or not.

[See in context](./src/services/ENV.js#L7-L13)



### Environment isolation

Per default, Whook takes the process environment as is
 but since it could lead to leaks when building for
 AWS Lambda or Google Cloud Functions one can isolate
 the process env when building.

[See in context](./src/services/ENV.js#L51-L56)



### `.env.NODE_ENV` files

You may need to keep some secrets out of your Git
 history. Whook uses `dotenv` to provide your such
 ability.

[See in context](./src/services/ENV.js#L64-L68)



## `

$autoload` service
The default Whook autoloader provides a simple way to
 load the constants, services and handlers of a Whook
 project.

[See in context](./src/services/_autoload.js#L11-L15)



### Configuration auto loading

Loading the configuration files is done according to the `NODE_ENV`
 environment variable. It basically requires a configuration hash
 where the keys are Knifecycle constants.

Let's load the configuration files as a convenient way
 to create constants on the fly

[See in context](./src/services/_autoload.js#L62-L69)



### Wrappers auto loading

We cannot inject the `WRAPPERS` in the auto loader when
 it is dynamically loaded so doing during the auto loader
 initialization if needed.

[See in context](./src/services/_autoload.js#L80-L84)



### API auto loading

We cannot inject the `API` in the auto loader since
 it is dynamically loaded so doing during the auto loader
 initialization.

[See in context](./src/services/_autoload.js#L98-L102)



### Constants

First of all the autoloader looks for constants in the
 previously loaded configuration.

[See in context](./src/services/_autoload.js#L139-L142)



### Handlers map

Here, we build the handlers map by injecting every handler required
 by the API.

[See in context](./src/services/_autoload.js#L151-L154)



### Service/handler loading

Finally, we either require the handler/service module if
 none of the previous strategies applyed.

[See in context](./src/services/_autoload.js#L181-L184)



## IP detection

If no `HOST` configuration is specified in dependencies nor in ENV,
 this service detects the machine host automagically.

[See in context](./src/services/HOST.js#L5-L8)



## Port detection

If no `PORT` configuration is specified in dependencies nor in ENV,
 this service detects a free port automagically.

[See in context](./src/services/PORT.js#L5-L8)

