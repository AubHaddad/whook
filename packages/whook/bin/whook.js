#! /usr/bin/env node

import path from 'path';
import runCLI from '../dist/cli.js';

const { prepareEnvironment } = await import(
  path.join(process.cwd(), 'dist', 'index.js')
);

await runCLI(prepareEnvironment);
