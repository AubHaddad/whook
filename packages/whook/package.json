{
  "metapak": {
    "configs": [
      "main",
      "readme",
      "tsesm",
      "eslint",
      "jest",
      "jsarch",
      "jsdocs"
    ],
    "data": {
      "childPackage": true,
      "files": "'src/**/*.ts'",
      "testsFiles": "'src/**/*.test.ts'",
      "distFiles": "'dist/**/*.js'",
      "ignore": [
        "dist"
      ],
      "bundleFiles": [
        "dist",
        "src"
      ]
    }
  },
  "name": "@whook/whook",
  "version": "13.2.1",
  "description": "Build strong and efficient REST web services.",
  "type": "module",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "bin": {
    "whook": "bin/whook.js"
  },
  "keywords": [
    "whook",
    "knifecycle",
    "REST",
    "HTTP",
    "OpenAPI",
    "web",
    "service",
    "handler",
    "server",
    "framework"
  ],
  "scripts": {
    "architecture": "jsarch 'src/**/*.ts' > ARCHITECTURE.md && git add ARCHITECTURE.md",
    "build": "rimraf 'dist' && tsc --outDir dist",
    "changelog": "conventional-changelog -p angular -i CHANGELOG.md -s",
    "cli": "env NODE_ENV=${NODE_ENV:-cli}",
    "cover": "npm run jest -- --coverage",
    "cz": "env NODE_ENV=${NODE_ENV:-cli} git cz",
    "doc": "echo \"# API\" > API.md; jsdoc2md 'dist/**/*.js' >> API.md && git add API.md",
    "jest": "NODE_OPTIONS=--experimental-vm-modules NODE_ENV=test jest",
    "lint": "eslint 'src/**/*.ts'",
    "metapak": "metapak",
    "prettier": "prettier --write 'src/**/*.ts'",
    "rebuild": "swc ./src -s -d dist -C jsc.target=es2022",
    "test": "npm run jest",
    "type-check": "tsc --pretty --noEmit"
  },
  "author": {
    "name": "Nicolas Froidure",
    "email": "nicolas.froidure@insertafter.com",
    "url": "http://insertafter.com/en/index.html"
  },
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/nfroidure/whook.git"
  },
  "bugs": {
    "url": "https://github.com/nfroidure/whook/issues"
  },
  "homepage": "https://github.com/nfroidure/whook",
  "peerDependencies": {
    "esbuild": ">=0.15.6",
    "esbuild-node-externals": "^1.5.0"
  },
  "devDependencies": {
    "@swc/cli": "^0.1.62",
    "@swc/core": "^1.3.94",
    "@swc/helpers": "^0.5.3",
    "@swc/jest": "^0.2.29",
    "@types/fs-extra": "^9.0.13",
    "@types/inquirer": "^9.0.3",
    "@types/mkdirp": "^1.0.2",
    "@types/streamtest": "^1.2.0",
    "@types/yargs-parser": "^21.0.0",
    "@typescript-eslint/eslint-plugin": "^6.8.0",
    "@typescript-eslint/parser": "^6.8.0",
    "esbuild-node-externals": "^1.5.0",
    "eslint": "^8.52.0",
    "eslint-config-prettier": "^9.0.0",
    "eslint-plugin-prettier": "^5.0.1",
    "jest": "^29.7.0",
    "jsarch": "^6.0.3",
    "jsdoc-to-markdown": "^8.0.0",
    "metapak": "^5.1.7",
    "metapak-nfroidure": "15.0.4",
    "prettier": "^3.0.3",
    "rimraf": "^5.0.5",
    "streamtest": "^2.0.0",
    "typescript": "^5.2.2"
  },
  "dependencies": {
    "@whook/http-router": "^13.2.0",
    "@whook/http-server": "^13.2.0",
    "@whook/http-transaction": "^13.2.0",
    "ajv": "^8.12.0",
    "ajv-formats": "^2.1.1",
    "application-services": "^4.0.0",
    "camelcase": "^8.0.0",
    "common-services": "^15.0.0",
    "debug": "^4.3.4",
    "dotenv": "^16.3.1",
    "fs-extra": "^11.1.1",
    "inquirer": "^9.2.10",
    "internal-ip": "^8.0.0",
    "knifecycle": "^17.0.1",
    "miniquery": "^1.1.2",
    "mkdirp": "^3.0.1",
    "openapi-types": "^12.1.3",
    "pkg-dir": "^7.0.0",
    "portfinder": "^1.0.32",
    "schema2dts": "^5.1.0",
    "type-fest": "^4.2.0",
    "yargs-parser": "^21.1.1",
    "yerror": "^8.0.0"
  },
  "contributors": [],
  "engines": {
    "node": ">=18.16.0"
  },
  "jest": {
    "coverageReporters": [
      "lcov",
      "html"
    ],
    "testPathIgnorePatterns": [
      "/node_modules/"
    ],
    "roots": [
      "<rootDir>/src"
    ],
    "testEnvironment": "node",
    "transform": {
      "^.+\\.tsx?$": [
        "@swc/jest",
        {}
      ]
    },
    "moduleNameMapper": {
      "#(.*)": "<rootDir>/../../node_modules/$1",
      "(.+)\\.js": "$1"
    },
    "extensionsToTreatAsEsm": [
      ".ts"
    ],
    "prettierPath": null
  },
  "eslintConfig": {
    "extends": [
      "eslint:recommended",
      "plugin:prettier/recommended",
      "plugin:@typescript-eslint/eslint-recommended",
      "plugin:@typescript-eslint/recommended"
    ],
    "parserOptions": {
      "ecmaVersion": 2018,
      "sourceType": "script",
      "modules": true
    },
    "env": {
      "es6": true,
      "node": true,
      "jest": true,
      "mocha": true
    },
    "plugins": [
      "prettier"
    ],
    "rules": {
      "prettier/prettier": "error"
    },
    "parser": "@typescript-eslint/parser",
    "ignorePatterns": [
      "*.d.ts"
    ]
  },
  "prettier": {
    "semi": true,
    "printWidth": 80,
    "singleQuote": true,
    "trailingComma": "all",
    "proseWrap": "always"
  },
  "files": [
    "dist",
    "src",
    "LICENSE",
    "README.md",
    "CHANGELOG.md"
  ],
  "jsarch": {
    "parserOptions": {
      "plugins": [
        "typescript"
      ]
    }
  }
}