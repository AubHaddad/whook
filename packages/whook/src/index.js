import Knifecycle, { constant } from 'knifecycle';
import debug from 'debug';
import {
  initLogService,
  initTimeService,
  initRandomService,
  initDelayService,
  initProcessService,
} from 'common-services';
import initHTTPRouter from '@whook/http-router';
import initErrorHandler from '@whook/http-router/dist/errorHandler';
import initHTTPTransaction from '@whook/http-transaction';
import initHTTPServer from '@whook/http-server';
import initPORT from './services/PORT';
import initHOST from './services/HOST';
import initAutoload from './services/_autoload';
import initENV from './services/ENV';
import initCONFIGS from './services/CONFIGS';
import initProjectDir from './services/PROJECT_DIR';

/* Architecture Note #1: Server run
Whook exposes a `runServer` function to programmatically spawn
 its server. It is intended to be reusable and injectable so
 that projects can override the whole `whook` default behavior.
*/
export async function runServer(
  prepareEnvironment,
  prepareServer,
  injectedNames = [],
) {
  try {
    const { ENV, log, $destroy, ...services } = await prepareServer(
      [...new Set([...injectedNames, 'ENV', 'log', '$destroy'])],
      await prepareEnvironment(),
    );

    if (ENV.DRY_RUN) {
      log('warning', 'ðŸŒµ - Dry run, shutting down now!');
      return $destroy();
    }

    return { ENV, log, $destroy, ...services };
  } catch (err) {
    // eslint-disable-next-line
    console.error('ðŸ’€ - Cannot launch the process:', err.stack);
    //process.exit(1);
  }
}

/* Architecture Note #2: Server preparation
Whook exposes a `prepareServer` function to create its server
 configuration. It takes eventually additional injections that
 would be required at a higher level and a
 [Knifecycle](https://github.com/nfroidure/knifecycle)
 containing the bootstrapped environment and allowing
 to complete and run the server.
*/
/**
 * Runs the Whook server
 * @param {Array<String>} injectedNames
 * Root dependencies names to instanciate and return
 * @param {Knifecycle} $
 * The Knifecycle instance to use for the server run
 * @returns Object
 * A promise of the injected services
 */
export async function prepareServer(injectedNames = [], $) {
  /* Architecture Note #2.1: Root injections
   * We need to inject `httpServer` and `process` to bring life to our
   *  server. We also inject `log` for logging purpose and custom other
   *  injected name that were required upfront.
   */
  const { log, ...services } = await $.run([
    ...new Set([...injectedNames, 'log', 'httpServer', 'process']),
  ]);

  log('warning', 'On air ðŸš€ðŸŒ•');

  return { log, ...services };
}

/* Architecture Note #3: Server environment
The Whook `prepareEnvironment` function aims to provide the complete
 server environment without effectively planning its run. It allows
 to use that environment for CLI or build purposes. It also
 provides a chance to override some services/constants
 before actually preparing the server.
 */
/**
 * Prepare the Whook server environment
 * @param {Knifecycle} $
 * The Knifecycle instance to set the various services
 * @returns Promise<Knifecycle>
 * A promise of the Knifecycle instance
 */
export async function prepareEnvironment($ = new Knifecycle()) {
  /* Architecture Note #3.1: `PWD` env var
  The Whook server heavily rely on the process working directory
   to dynamically load contents. We are making it available to
   the DI system as a constant.
   */
  const PWD = process.cwd();
  $.register(constant('PWD', PWD));

  /* Architecture Note #3.2: `NODE_ENV` env var
  Whook has different behaviors depending on the `NODE_ENV` value
   consider setting it to production before shipping.
   */
  const NODE_ENV = process.env.NODE_ENV || 'development';
  $.register(constant('NODE_ENV', NODE_ENV));

  /* Architecture Note #3.3: `WHOOK_PLUGINS` and `PROJECT_SRC`
  Whook need to know where to look up for things like
   commands / handlers etc...
   */
  $.register(constant('WHOOK_PLUGINS', ['@whook/whook']));

  /* Architecture Note #3.4: Logging
  Whook's default logger write to the NodeJS default console
   except for debugging messages where it uses the `debug`
   module so that you can set the `DEBUG` environment
   variable to `whook` and get debug messages in output.
   */
  $.register(constant('debug', debug('whook')));
  $.register(
    constant('logger', {
      // eslint-disable-next-line
      error: console.error.bind(console),
      // eslint-disable-next-line
      info: console.info.bind(console),
      // eslint-disable-next-line
      warning: console.error.bind(console),
    }),
  );
  $.register(constant('exit', process.exit));

  /* Architecture Note #3.5: Initializers
  Whook's embed a few default initializers proxied from
   `common-services`, `@whook/http-router` or its own
   `src/services` folder. It can be wrapped or overriden,
   at will, later in project's main file.
   */
  [
    initLogService,
    initTimeService,
    initRandomService,
    initDelayService,
    initProcessService,
    initHTTPRouter,
    initHTTPTransaction,
    initHTTPServer,
    initErrorHandler,
    initPORT,
    initHOST,
    initENV,
    initCONFIGS,
    initAutoload,
    initProjectDir,
  ].forEach($.register.bind($));

  return $;
}
