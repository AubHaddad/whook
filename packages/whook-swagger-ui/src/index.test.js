import { constant, initializer } from 'knifecycle';
import axios from 'axios';
import { prepareServer, prepareEnvironment } from 'whook';
import { initHTTPRouter } from 'swagger-http-router';
import wrapHTTPRouterWithSwaggerUI from '.';

describe('wrapHTTPRouterWithSwaggerUI', () => {
  const PORT = 8888;
  const logger = {
    info: jest.fn(),
    error: jest.fn(),
  };
  const debug = jest.fn();
  const time = jest.fn();
  const $autoload = jest.fn(async () => ({}.undef));
  let $;

  beforeEach(() => {
    logger.info.mockReset();
    logger.error.mockReset();
    debug.mockReset();
    time.mockReset();
    $autoload.mockClear();
  });

  beforeEach(async () => {
    $ = await prepareEnvironment();

    $.register(
      initializer(
        {
          name: '$autoload',
          type: 'service',
          options: { singleton: true },
        },
        async () => $autoload,
      ),
    );
    $.register(
      constant('API', {
        host: `localhost:${PORT}`,
        swagger: '2.0',
        info: {
          version: '1.0.0',
          title: 'Sample Swagger',
          description: 'A sample Swagger file for testing purpose.',
        },
        basePath: '/v1',
        schemes: ['http'],
        paths: {
          '/ping': {
            get: {
              operationId: 'getPing',
              summary: "Checks API's availability.",
              consumes: ['application/json'],
              produces: ['application/json'],
              responses: {
                '200': {
                  description: 'Pong',
                },
              },
            },
          },
        },
      }),
    );
    $.register(constant('ENV', {}));
    $.register(constant('NODE_ENV', 'test'));
    $.register(constant('HOST', 'localhost'));
    $.register(constant('WRAPPERS', []));
    $.register(
      constant('HANDLERS', {
        getPing: jest.fn(async () => ({
          status: 200,
          headers: { 'content-type': 'application/json' },
          body: { ping: 'pong' },
        })),
      }),
    );
    $.register(constant('logger', logger));
    $.register(constant('debug', debug));
    $.register(constant('time', time));
    $.register(constant('NODE_ENVS', ['test']));
  });

  it('should work', async () => {
    $.register(constant('PORT', PORT));
    $.register(wrapHTTPRouterWithSwaggerUI(initHTTPRouter));
    $.register(
      constant('CONFIG', {
        basePath: '/v1',
      }),
    );
    $.register(constant('NODE_ENV', 'test'));
    $.register(constant('DEBUG_NODE_ENVS', ['test']));

    time.mockReturnValue(new Date('2010-03-06T00:00:00Z').getTime());

    const { $destroy } = await prepareServer(
      ['$destroy', 'httpServer', 'process'],
      $,
    );
    const { status, headers, data } = await axios.get(
      `http://localhost:${PORT}/v1/ping`,
    );

    await $destroy();

    expect({
      status,
      headers: {
        ...headers,
        // Erasing the Date header that may be added by Axios :/
        date: {}.undef,
      },
      data,

      debugCalls: debug.mock.calls,
      logInfoCalls: logger.info.mock.calls,
      logErrorCalls: logger.error.mock.calls,
      autoloaderCalls: $autoload.mock.calls,
    }).toMatchSnapshot();
  });

  it('should serve Swagger HTML', async () => {
    $.register(constant('PORT', PORT + 2));
    $.register(wrapHTTPRouterWithSwaggerUI(initHTTPRouter));
    $.register(
      constant('CONFIG', {
        localURL: `http://localhost:${PORT + 2}`,
        basePath: '/v1',
      }),
    );
    $.register(constant('NODE_ENV', 'test'));
    $.register(constant('DEBUG_NODE_ENVS', ['test']));

    time.mockReturnValue(new Date('2010-03-06T00:00:00Z').getTime());

    const { $destroy } = await prepareServer(
      ['$destroy', 'httpServer', 'process'],
      $,
    );
    const { status, headers, data } = await axios.get(
      `http://localhost:${PORT + 2}/docs`,
    );

    await $destroy();

    expect({
      status,
      headers: {
        ...headers,
        // Erasing the Date header that may be added by Axios :/
        date: {}.undef,
        etag: {}.undef,
        'last-modified': {}.undef,
      },
      data,

      debugCalls: debug.mock.calls,
      logInfoCalls: logger.info.mock.calls,
      logErrorCalls: logger.error.mock.calls,
      autoloaderCalls: $autoload.mock.calls,
    }).toMatchSnapshot();
  });

  it('should be bypassed with no debug env', async () => {
    $.register(constant('PORT', PORT + 1));
    $.register(wrapHTTPRouterWithSwaggerUI(initHTTPRouter));
    $.register(
      constant('CONFIG', {
        localURL: `http://localhost:${PORT + 1}`,
        basePath: '/v1',
      }),
    );
    $.register(constant('NODE_ENV', 'test'));
    $.register(constant('DEBUG_NODE_ENVS', []));

    time.mockReturnValue(new Date('2012-01-15T00:00:00Z').getTime());

    const { $destroy } = await prepareServer(
      ['$destroy', 'httpServer', 'process'],
      $,
    );

    await $destroy();

    expect({
      debugCalls: debug.mock.calls,
      logInfoCalls: logger.info.mock.calls,
      logErrorCalls: logger.error.mock.calls,
      autoloaderCalls: $autoload.mock.calls,
    }).toMatchSnapshot();
  });
});
