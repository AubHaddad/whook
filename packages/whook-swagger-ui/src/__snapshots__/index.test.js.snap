// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapHTTPRouterWithSwaggerUI should be bypassed with no debug env 1`] = `
Object {
  "autoloaderCalls": Array [
    Array [
      "MAX_HEADERS_COUNT",
    ],
    Array [
      "KEEP_ALIVE_TIMEOUT",
    ],
    Array [
      "MAX_CONNECTIONS",
    ],
    Array [
      "TIMEOUT",
    ],
    Array [
      "PROCESS_NAME",
    ],
    Array [
      "SIGNALS",
    ],
    Array [
      "BUFFER_LIMIT",
    ],
    Array [
      "PARSERS",
    ],
    Array [
      "STRINGIFYERS",
    ],
    Array [
      "DECODERS",
    ],
    Array [
      "ENCODERS",
    ],
    Array [
      "QUERY_PARSER",
    ],
    Array [
      "TRANSACTIONS",
    ],
    Array [
      "uniqueId",
    ],
  ],
  "debugCalls": Array [
    Array [
      "Logging service initialized.",
    ],
    Array [
      "Time service initialized.",
    ],
    Array [
      "Delay service initialized.",
    ],
    Array [
      "Running in \\"test\\" environment.",
    ],
    Array [
      "Process service initialized.",
    ],
    Array [
      "HTTP Transaction initialized.",
    ],
    Array [
      "HTTP Router initialized.",
    ],
    Array [
      "Closing HTTP server.",
    ],
    Array [
      "HTTP server closed",
    ],
    Array [
      "Cancelling pending timeouts:",
      0,
    ],
  ],
  "logErrorCalls": Array [],
  "logInfoCalls": Array [
    Array [
      "HTTP Server listening at \\"http://localhost:8888\\".",
    ],
    Array [
      "On air ðŸš€ðŸŒ•",
    ],
  ],
}
`;

exports[`wrapHTTPRouterWithSwaggerUI should work 1`] = `
Object {
  "autoloaderCalls": Array [
    Array [
      "MAX_HEADERS_COUNT",
    ],
    Array [
      "KEEP_ALIVE_TIMEOUT",
    ],
    Array [
      "MAX_CONNECTIONS",
    ],
    Array [
      "TIMEOUT",
    ],
    Array [
      "PROCESS_NAME",
    ],
    Array [
      "SIGNALS",
    ],
    Array [
      "BUFFER_LIMIT",
    ],
    Array [
      "PARSERS",
    ],
    Array [
      "STRINGIFYERS",
    ],
    Array [
      "DECODERS",
    ],
    Array [
      "ENCODERS",
    ],
    Array [
      "QUERY_PARSER",
    ],
    Array [
      "TRANSACTIONS",
    ],
    Array [
      "uniqueId",
    ],
  ],
  "debugCalls": Array [
    Array [
      "Logging service initialized.",
    ],
    Array [
      "Time service initialized.",
    ],
    Array [
      "Delay service initialized.",
    ],
    Array [
      "Running in \\"test\\" environment.",
    ],
    Array [
      "Process service initialized.",
    ],
    Array [
      "HTTP Transaction initialized.",
    ],
    Array [
      "HTTP Router initialized.",
    ],
    Array [
      "Closing HTTP server.",
    ],
    Array [
      "HTTP server closed",
    ],
    Array [
      "Cancelling pending timeouts:",
      0,
    ],
  ],
  "logErrorCalls": Array [],
  "logInfoCalls": Array [
    Array [
      "Serving the API docs: http://localhost:888/docs?url=http%3A%2F%2Flocalhost%3A888%2Fv1%2FopenAPI",
    ],
    Array [
      "HTTP Server listening at \\"http://localhost:8888\\".",
    ],
    Array [
      "On air ðŸš€ðŸŒ•",
    ],
  ],
}
`;
